File: 108-convert-sorted-array-to-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
    if len(nums) == 0{
        return nil
    }
    return makeNode(nums, 0, len(nums)-1)
}

func makeNode(nums[] int, left, right int) *TreeNode{
    if right < left{
        return nil
    }
    mid := (left + right) / 2
    nd := &TreeNode{nums[mid], nil, nil}
    nd.Left = makeNode(nums, left, mid-1)
    nd.Right = makeNode(nums, mid+1, right)
    return nd
}

File: bwu/binary_search/108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        n = len(nums)
        if not n:
            return None
            
        mid = nums[n / 2]
        node = TreeNode(mid)
        left = self.sortedArrayToBST(nums[: n / 2])
        right = self.sortedArrayToBST(nums[(n / 2) + 1:])
        node.left = left
        node.right = right
        
        return node
